dra-evolution/pkg/api/capacity_types.go

```diff
@@ -59,14 +59,13 @@ type ResourcePoolSpec struct {
 	// vendor of the driver.
 	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
 
-	// SharedCapacity defines the set of shared capacity consumable by
-	// devices in this ResourceSlice.
+	// SliceContent represents the type of data in this slice. Currently the only
+	// valid values is `Devices`. As we add features, new values will be added
+	// to this discriminator. This allows older schedulers to know if they should
+	// ignore the slice, due to it being of a type they do not understand.
 	//
-	// Must not have more than 128 entries.
-	//
-	// +listType=atomic
-	// +optional
-	SharedCapacity []SharedCapacity `json:"sharedCapacity,omitempty"`
+	// +required
+	SliceContent string `json:"sliceContent"`
```

Comment:
Should this be an enum instead, @pohly ?

Reply:
So the scheduler will decode this object, read this field, and then decide if it knows how to work with it or not? an enum seems appropriate then, since this should only be allowed to have a specific set of values.

Reply:
Do we need this?

If a current-rev scheduler only knows how to work with "devices", it uses the `slice.devices` field to find those.

If a future API adds "gizmos" or some variant of "devices" (remember, I wanted to keep the current field named as "namedDevices"...), then it adds a `slice.gizmos` field. A current-rev scheduler will see an empty slice and ignore it because it contains no devices.

Reply:
Or is the intent to add additional fields to the `Device` struct? In that case such a `SliceContent` discriminator would be useful because an unknown value tells old clients that there is something in the `slice.devices` that they don't know about, but should know.

I just wouldn't call it `SliceContent`, but rather `DeviceType`: it's a description of the `slice.devices`.

If we add `slice.gizmos` (separate field), then they can have a `slice.gizmoType`.

Reply:
We likely would add another field in addition to "Devices" to qualify or add onto the devices. 

If we want to avoid the discriminator, I would avoid the term "Devices". But I don't think "NamedDevices" works because "named" isn't a differentiating feature. Maybe "SimpleDevices" or something?

If we keep the discriminatory, I do think "DeviceType" is right - because it wouldn't make sense for "gizmos", for example.

Reply:
I like the discriminator and don't feel strongly about what it is called (though I'd like it to have some symmetry with other fields -- what did we decide to call the replacement for `PoolDeviceCount` now that we are going with a `SliceCount` instead)?

I think Patrick is suggesting that field itself becomes a one-of (with a future `GizmoType` field, for example), such that it pairs with the inclusion of a `devices` list (or a `gizmos` list in the future).

Reply:
My point was that there is no need for one-ofs. A slice could contain both `SimpleDevices []SimpleDevice` and `ComplexDevices []ComplexDevice` at the same time. Depending what a scheduler supports, it can allocate from `SimpleDevices` or from both.

However, for the maximum size calculation it might make sense to require that either one or the other slice is empty. That doesn't change my argument, a scheduler which only supports `SimpleDevices []SimpleDevice` will see an empty slice and safely ignore it.

Reply:
Having said that, I am okay with adding the discriminator. It may be redundant when adding a second slice with a different type, but for the case where the same type gets extended (https://github.com/kubernetes-sigs/wg-device-management/pull/36/files/3a7aa4f9789c7979426ea3c65497545136428f55#r1646662809) it is useful.

My question about validation after a downgrade applies also here.

Reply:
According to the bullet on "enumerated types" [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas), we should be OK as long as we document that schedulers that see unknown slice types should ignore those slices. I think older clients doing a *write* would be subject to the API validation of the API they are using, meaning they would not be able to use the new enum. I think if there is a field associated with the new enum value, we should clear it (the client doesn't know about the new enum, but we do). I *think* that's consistent with the logic described [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version), but we should discuss with api reviewers. 

One question: Should (can?) we ensure all slices of a pool are of the same slice type? If not, old schedulers would have to ignore the whole pool, at least for "all", if ANY slice is of an unknown type...

Reply:
It would seem to me that adding an enum value should require a bump in the API version.

Reply:
Once we have reached v1, we cannot bump it any further?

I'm still not exactly clear on how to implement validation, but there must have been precedences, so I guess it is okay.

Reply:
A discriminator is *required* when you want the absence of config to imply a default, so that version skew isn't perceived as the default case.

For cases that are truly one-of (exactly one field, empty lists don't count), then a discriminator is not needed, because any actor can perceive the lack of a value, and choose not to act.

Whether we extend `Device` to encompass partitions (assuming we can do that safely) or we add `PartitionableDevice` here, it shouldn't matter -- both are viable.  Likewise, whether we want devices/gizmos/widgets to co-exist in a slice or to be mutually exclusive, I think both are viable?  There may be *other* things to consider to make those decisions, but I don't think *this* aspect matters.